flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/TaskExecutorResourceUtils.java

```diff
@@ -0,0 +1,333 @@
+		checkArgument(totalFlinkExternalMemorySize.getBytes() < totalProcessMemorySize.getBytes(),
+			"Sum of configured JVM Metaspace (" + jvmMetaspaceSize.toString()
+				+ ") and JVM Overhead (" + jvmOverheadSize.toString()
+				+ ") exceed configured Total Process memory (" + totalProcessMemorySize.toString() + ").");
+		final MemorySize totalFlinkMemorySize = totalProcessMemorySize.subtract(totalFlinkExternalMemorySize);
+
+		// derive total flink internal memory sizes from derived total flink memory size
+		final Tuple6<MemorySize, MemorySize, MemorySize, MemorySize, MemorySize, MemorySize> totalFLinkInternalMemorySizeTuple6 =
+			deriveTotalFlinkInternalMemorySizes(config, totalFlinkMemorySize);
+
+		return new TaskExecutorResourceSpec(
+			totalFLinkInternalMemorySizeTuple6.f0, // frameworkHeapMemorySize
+			totalFLinkInternalMemorySizeTuple6.f1, // taskHeapMemorySize
+			totalFLinkInternalMemorySizeTuple6.f2, // taskOffHeapMemorySize
+			totalFLinkInternalMemorySizeTuple6.f3, // shuffleMemorySize
+			totalFLinkInternalMemorySizeTuple6.f4, // onHeapManagedMemorySize
+			totalFLinkInternalMemorySizeTuple6.f5, // offHeapManagedMemorySize
+			jvmMetaspaceSize,
+			jvmOverheadSize);
+	}
+
+	private static Tuple2<MemorySize, MemorySize> deriveTotalFlinkExternalMemorySizes(
+		final Configuration config, final MemorySize totalFlinkMemorySize) {
+		final MemorySize jvmMetaspaceSize = getJvmMetaspaceSize(config);
+		final MemorySize jvmOverheadSize = deriveJvmOverheadWithInverseFraction(config,
+			totalFlinkMemorySize.add(jvmMetaspaceSize));
+		return new Tuple2<>(jvmMetaspaceSize, jvmOverheadSize);
+	}
+
+	private static Tuple6<MemorySize, MemorySize, MemorySize, MemorySize, MemorySize, MemorySize> deriveTotalFlinkInternalMemorySizes(
+		final Configuration config, final MemorySize totalFlinkMemorySize) {
+		final MemorySize frameworkHeapMemorySize = getFrameworkHeapMemorySize(config);
+		final MemorySize taskOffHeapMemorySize = getTaskOffHeapMemorySize(config);
+
+		final MemorySize shuffleMemorySize = deriveShuffleMemoryWithFraction(config, totalFlinkMemorySize);
+		final MemorySize managedMemorySize = deriveManagedMemoryAbsoluteOrWithFraction(config, totalFlinkMemorySize);
+
+		final MemorySize totalFlinkExceptTaskHeapMemorySize =
+			frameworkHeapMemorySize.add(taskOffHeapMemorySize).add(shuffleMemorySize).add(managedMemorySize);
+		checkArgument(totalFlinkExceptTaskHeapMemorySize.getBytes() < totalFlinkMemorySize.getBytes(),
+			"Sum of configured Framework Heap Memory (" + frameworkHeapMemorySize.toString()
+				+ "), Task Off-Heap Memory (" + taskOffHeapMemorySize.toString()
+				+ "), Shuffle Memory (" + shuffleMemorySize.toString()
+				+ ") and Managed Memory (" + managedMemorySize.toString()
+				+ ") exceed configured Total Flink Memory (" + totalFlinkMemorySize.toString() + ").");
+		final MemorySize taskHeapMemorySize = totalFlinkMemorySize.subtract(totalFlinkExceptTaskHeapMemorySize);
```

Comment:
Currently in this PR, `taskmanager.memory.task.heap.size` will be ignored in such cases.

I also noticed your concern during the implementation, that the user explicitly configured task heap memory size is not taking effect only because managed memory size is not explicitly configured at the same time, which is quite not intuitive to the users.

However, if we want to respect user's explicit configurations as much as possible, then it would be a question that which implicit configuration should be overwritten. E.g., if the user explicitly configured total flink memory size and task heap memory size, and the explicitly configured task heap memory size is larger than derived from total flink memory size, the which memory pool should we shrink to guarantee the explicit configured task heap memory size? Leaving alone task off heap memory and framework heap memory (which both have default absolute values), should we prefer shrinking managed memory or shuffle memory? And do we respect the implicit min sizes of managed/shuffle memory? What if shrinking either of managed/shuffle memory can not give enough memory to the explicitly configured task heap memory (wrt the min sizes), so that we need to shrink both of them? How many should we shrink on each of them?

It seems to me that the calculation logics can be complicated if we try to respect all explicit user configurations, which may also not good for user to understand and for maintainability. Thus before we decide how to deal with this tradeoff, I implemented this PR strictly following FLIP-49 design doc.

Reply:
I think the logic should be to try to minimize other pool budgets if the minimum configured value allows it and if this is not enough, then we fail. Which pool is shrunk, I would not give a guarantee. It should also be ok to pick an arbitrary one because otherwise the user would have configured a different min value.